source/agora/common/Transaction.d

```diff
@@ -0,0 +1,85 @@
+/*******************************************************************************
+
+    Defines the data structure of a transaction
+
+    The current design is heavily influenced by Bitcoin: as we have an UTXO,
+    starting with a simple input/output approach seems to make the most sense.
+    Script is not implemented: instead, we currently only do a simple signature
+    verification.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Transaction;
+
+import agora.common.Data;
+import agora.common.crypto.Key;
+
+
+/*******************************************************************************
+
+    Currency amount to use
+
+    This is simply an alias to `long` at the moment, and should be made smarter.
+    Currently this has two major drawbacks:
+    - It does not do any overflow / underflow checks (over the total amount)
+    - It can go negative
+    We should probably wrap this in a smarter type to do currency operations,
+    and move it to a shared location, as currency operations might be performed
+    independently of `Transaction`.
+
+*******************************************************************************/
+
+public alias Amount = long;
```

Comment:
Bitcoin seems to allow negative values (at least internally?), but I'm not sure why. Do you know?

Reply:
Not exactly. I imagine that might be related to dealing with `a - b` (remember Don hating on unsigned types ? ;) )

Reply:
Ah that's a very good point. I remember the implicit `ulong` => `uint` conversion in D1 as well. That was a fun deployment.